tests/test_pca.py

```diff
@@ -739,3 +741,64 @@ def test_fractional_n_components():
         pca.fit(X)
 
     assert w.match("Fractional 'n_components'")
+
+
+@pytest.mark.parametrize("solver", ["auto", "tsqr", "randomized", "full"])
+@pytest.mark.parametrize("fn", ["fit", "fit_transform"])
+@pytest.mark.parametrize("errors", ["raise", "warn"])
+def test_unknown_shapes(fn, solver, errors):
+    df = pd.DataFrame({"0": [0, 0, 1, 1], "1": [0, 0, 1, 1], "2": [2, 4, 5, 2]})
+    ddf = dask.dataframe.from_pandas(df, npartitions=2)
+
+    pca = dd.PCA(n_components=2, svd_solver=solver, errors=errors)
+    fit_fn = getattr(pca, fn)
+    X = ddf.values
+    assert np.isnan(X.shape[0])
+
+    match = "The check on n_components can't be completed"
+    if solver == "auto":
+        with pytest.raises(ValueError, match="automatically choose PCA solver"):
+            fit_fn(X)
+    elif errors == "raise":
+        with pytest.raises(ValueError, match=match):
+            fit_fn(X)
+    elif errors == "warn":
+        with pytest.warns(UserWarning, match=match):
+            X_hat = fit_fn(X)
+        assert hasattr(pca, "components_")
+        assert pca.n_components_ == 2
+        assert pca.n_features_ == 3
+        assert np.isnan(pca.n_samples_)
+        if fn == "fit_transform":
+            assert np.isnan(X_hat.shape[0])
+            assert X_hat.shape[1] == 2
+
+
+@pytest.mark.parametrize("solver", ["randomized", "tsqr", "full"])
+def test_dataframe_pca_fat_shape(solver):
+    X = np.random.randn(2, 10)
+    df = pd.DataFrame(X)
+    ddf = dask.dataframe.from_pandas(df, npartitions=2)
+    X = ddf.values
+    assert np.isnan(X.shape[0])
+
+    pca = dd.PCA(n_components=3, svd_solver=solver, errors="warn")
+    with pytest.warns(
+        UserWarning, match="The check on n_components can't be completed"
+    ):
+        if solver == "randomized":
+            # This test encodes the fact that PCA will fail silently with the
+            # shapes!
+            X_hat = pca.fit_transform(X).compute()
+            assert X_hat.shape == (2, 2) != (2, 3)
```

Comment:
This behavior surprised me too.

This test is for when the user specifies `n_components` to be too large, and making sure it's caught. When `solver == "randomized"`, PCA will silently produce an array of the *correct* size `(2, 2)` (which is not what the user specified `(2, 3)`).

I'd rather not code the fact that Dask-ML ignores user-input into a test. I skip the `assert` statements and have a comment explaining what happens:

``` python
        if solver == "randomized":
            # This condition will "fail" silently! It'll produce an array of the
            # *mathematically correct* number of columns (2, not 3), even
            # though the user specified n_components=3 (n_components is larger
            # than 2, the number of singular values)
            pass
```

Reply:
> PCA will silently produce an array of the correct size (2, 2)

Why do you say (2, 2) is the correct size? It's the size of the minimum of `(n_samples, n_features)`, but it's not what the user requested.

At the end of `.fit()` can we check if the number of components doesn't match the user-specified `n_components` (when the user specifies an integer).

Reply:
> Why do you say (2, 2) is the correct size? It's the size of the minimum of (n_samples, n_features), but it's not what the user requested.

(2, 2) is mathematically correct. (2, 3) follows the user-specification (even though PCA can not produce it).

> At the end of .fit() can we check if the number of components doesn't match the user-specified n_components (when the user specifies an integer).

That's a good idea. I've implemented that change.

Reply:
So is there any reason to keep `errors="warn"` or "ignore" anymore? This seems like we can achieve the correct behavior all the time.

The only difference is that with known shapes we figure it out before fitting, while with unknown shapes we figure it out after?

Reply:
> The only difference is that with known shapes we figure it out before fitting, while with unknown shapes we figure it out after?

Yup, almost. PCA works fine if `n_components` is properly specified.

I think it'd be useful to have `errors` to make the user understand they're doing something risky and disabling some checks; an uninformative error message is raised otherwise if `n_components` is too large ("operands could not be broadcast together").

Reply:
When is this raised? Do you have an example?

Reply:
> When is this raised? Do you have an example?

In short, `errors` is used to protect against when both cases below are true:

- [ ] the user specifies `n_components > X.shape[0]`
- [ ] `solver != 'randomized'`.

The `solver == 'randomized'` works, but it will ignore the user-specified `n_components`. PCA will issue a warning.

Here are the two conditions for `n_components`:

1. `n_components <= X.shape[0]`:
    * `errors` is used to warn the check can't be completed. To continue, `errors == 'warn'`.
    * the SVD finishes succesfully in `_fit`
2. `n_components > X.shape[0]`
    * `errors` is used to warn the check can't be completed. To continue, `errors == 'warn'`.
    * Two cases:
        1. if `solver == 'randomized'`, PCA successfully finishes and continues with `pca.n_components_ = 2` (and warns the user).
        2. if `solver != 'randomized'`, the SVD in `_fit` errors on "operands could not be broadcast together".

There's no way to tell is the SVD will fail ahead of time.

[test_unknown_shapes]:https://github.com/dask/dask-ml/pull/543/files#diff-7f86330d39ca24bd1ab2bfd063eb893bR749

Reply:
...or I could wrap the SVD computation in a try/except block. We know when the error happens, and it's only in some cases.

Reply:
> There's no way to tell is the SVD will fail ahead of time.

Understood. I'm happy to raise after the fit when `X.shape[0]` is unknown.